[zhiayang] iP

zhiayang opened the pull request.

Commit:
Add Gradle support

Commit:
First commit; Implement level 1

Commit:
Check for Scanner EOF before reading

Commit:
Implement level 2

Commit:
Add build script to gitignore

Commit:
Add Task class; implement level 3

Commit:
Refactor command parsing

Commit:
"Fix" tabs to spaces

Commit:
Re-refactor command code; implement level 4

Commit:
refactor printing again

Commit:
Add test script, remove old stuff

Commit:
Simplify Task::toString() and derivatives

Commit:
Fix NaN error for task completion percentage

Commit:
Remove unused variable

Commit:
Refactor command processing, implement error handling

Commit:
Implement task deletion (level 6)

Commit:
Update tests

Commit:
Remove exception name from toString

Commit:
Fix parsing for Deadline and Event tasks

Also, add tests to cover the missing corner cases.

Commit:
Refactor duplicate code for task parsing

Create a new helper class DatedTask that consolidates the
code for parsing dates for Event and Deadline.

Commit:
Add getters and setters to the Bot

Commit:
Add getters for Event's and Deadline's respective date fields

Commit:
Add Either utility class

Commit:
Create InvalidDatabaseException class

Commit:
Add the Database class, and allow loading Tasks from disk

Commit:
Add functionality to save Tasks to disk

Commit:
Add support for date parsing to DatedTask and Deadline

Commit:
Change the printed format of dates as required

Commit:
Replace formatting on Event to match Deadline

Commit:
Add todo

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Add 'reset' command to clear the task list

Mainly used for testing to start with a blank slate.

Commit:
Update Database to handle parsing dates from disk

Commit:
Explicitly set the output format for the date when saving

Commit:
Update test cases to reflect new date input

Commit:
Encapsulate task list functionality into a TaskList class

Also, create a Frontend class that serves very little purpose.

Commit:
Save the list of tasks after every command

Commit:
Move the user-input functionality into the Frontend

Commit:
Refactor into packages

Commit:
Allow changing the input and output streams of Frontend

Commit:
Add missing operator== overload for Tasks

Commit:
Add JUnit tests for TaskList, and MainTest

MainTest replaces the text-based input test, by reading the same
files and doing a comparison.

Commit:
Add script to make jar

Commit:
Implement find functionality

Commit:
Add mapLeft/mapRight methods to Either

Commit:
Add javadocs to all public methods

Commit:
Merge branch 'branch-Level-9' into master

Commit:
Merge branch 'branch-A-JavaDoc' into master

Commit:
Follow the coding standard

Commit:
Merge branch 'branch-A-CodingStandard' into HEAD

Commit:
Update javadocs for new Either methods

Commit:
build.gradle: Update version to 8.29

kunnan97 commented on src/main/java/ikura/Bot.java:
```diff
@@ -0,0 +1,234 @@
+     */
+    public Bot(Frontend ui, TaskList tasks) {
+        this.tasks  = tasks;
+        this.ui     = ui;
```
I believe operators(e.g. =) should be surrounded by 1 space? Seems like this happens at other places too. Could be because I am referring to Java's coding standard. Not sure if Kotlin follows that too.

zhiayang replied:
it's not covered in the 2103 guide, but the google style guide they base it on says that horizontal alignment is [acceptable](https://google.github.io/styleguide/javaguide.html#s4.6.3-horizontal-alignment).

kunnan97 replied:
Alright then all is good!

kunnan97 commented on src/main/java/ikura/task/DatedTask.java:
```diff
@@ -0,0 +1,74 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+import ikura.util.Pair;
```
Packages imported should group in a better way.

zhiayang replied:
hmm, how would you suggest I group them? my current convention is java classes first, followed by project classes, then exception types.

afaik the coding style doesn't prescribe a specific ordering, right?

kunnan97 replied:
oh oops I think I got confused between the java and ikura packages, ignore this please.

kunnan97 commented on src/main/java/ikura/task/Event.java:
```diff
@@ -0,0 +1,76 @@
+     * @param deadline the date of this event.
+     */
+    public Event(String name, LocalDate date) {
+        super(name);
+		this.eventDate = date;
```
Some spacing issues here

zhiayang replied:
ah yes, my editor is configured to tabs by default, i must've missed this one. thanks!

kunnan97 reviewed (commented):
Overall, looks good to me, just some styling issues.

Commit:
Fix style errors

Commit:
Add checkstyle configurations

Commit:
Merge branch 'branch-A-Gradle' into master

This adds gradle build support.

Commit:
Move Main class to ikura package

Commit:
Update gradle build file

Commit:
Appease the checkstyle

Commit:
Rename Task::getName to getTitle

Tasks now have both a title and a description; what used to be
the "name" is now the title, and the description is empty by
default.

Commit:
Add constructor for Task taking a description

Commit:
Preliminary implementation of Gui SKELETON

It's really just a skeleton ok

AaronnSeah commented on src/main/java/ikura/Bot.java:
```diff
@@ -0,0 +1,234 @@
+
+        this.ui.beginLog();
+
+        Optional.ofNullable(Map.<String, BiConsumer<String, String>>of(
```
Perhaps it may be good to set the Map object as an attribute of the Bot class?

zhiayang replied:
good point, since it's constant.

AaronnSeah commented on src/main/java/ikura/Bot.java:
```diff
@@ -0,0 +1,234 @@
+        this.ui.beginLog();
+
+        Optional.ofNullable(Map.<String, BiConsumer<String, String>>of(
+            "done",     this::cmdDone,
```
Perhaps this code could be simplified for readability?

zhiayang replied:
how would you suggest to simplify it? to me, it's a lot more clear than a series of switch/cases or even if-else blocks.

AaronnSeah replied:
One suggestion I can think of is to make the map an attribute, and abstract out the lambda as a method with a meaningful method name. That way the code would be much shorter and can be understood more easily and more quickly.

AaronnSeah commented on src/main/java/ikura/Bot.java:
```diff
@@ -0,0 +1,234 @@
+        this.ui.endLog();
+
+        // save the tasks every time, i suppose.
+        this.tasks.save();
```
Perhaps the comment is not necessary as the code and comment indicates the same meaning?

AaronnSeah commented on src/main/java/ikura/Bot.java:
```diff
@@ -0,0 +1,234 @@
+            .forEach(x -> this.ui.println("%s", x));
+    }
+
+    private void cmdDelete(String cmd, String input) {
```
Perhaps the naming for the method could be better, such as handleDeleteCommand?

AaronnSeah commented on src/main/java/ikura/Database.java:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    /**
+     * Load the list of tasks from the database.
```
Perhaps you could use Loads instead of Load to be consistent with the coding standard?

zhiayang replied:
are you referring to the javadoc? in that case ya good point

AaronnSeah commented on src/main/java/ikura/Database.java:
```diff
@@ -0,0 +1,242 @@
+     * @throws IOException if there was an error reading the file.
+     * @throws InvalidDatabaseException if there were malformed lines in the file.
+     */
+    public List<Task> loadTasks() throws IOException, InvalidDatabaseException {
```
Perhaps the method can be <= 30 lines of code to be consistent with the code quality standard "avoid long methods"?

AaronnSeah commented on src/main/java/ikura/Database.java:
```diff
@@ -0,0 +1,242 @@
+                        }
+                    } else {
+                        return Either.left(String.format("invalid type '%c' (expected one of 'T', 'D', 'E')", type));
+                    }
```
Perhaps there could be less deep nesting to be consistent with the code quality standard of "Avoid Deep Nesting"

zhiayang replied:
you're right, i should refactor this method eventually.

AaronnSeah commented on src/main/java/ikura/Main.java:
```diff
@@ -0,0 +1,28 @@
+
+        ui.greet();
+        while (ui.readLine().map(bot::processCommand).orElse(false))
+            ;
```
Perhaps the ; could be in line 25?

zhiayang replied:
tbh i think the style guide would say that i just make an empty block with `{ }`.

anyway (style guide aside) my motivation for putting the semicolon on a separate line for empty loop bodies is that it makes it explicit that the body is empty, because it can be easy to miss the semicolon at the end of the line and think that the proceeding statement is part of the loop.

AaronnSeah commented on src/main/java/ikura/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+     * @return the Task if it exists, or an empty optional otherwise.
+     */
+    public Optional<Task> getTaskByNumber(int number) {
+        // asdf
```
Perhaps this comment is not needed?

AaronnSeah reviewed (commented):
LGTM, just a few nits to pick.

Commit:
Change left side to a ListView

Commit:
Flesh out the gui further; layout is probably final now

Commit:
Refactor Tasks to actually contain descriptions

Database version updated from 1 to 2.

Commit:
Merge branch 'master' into branch-Level-10

Commit:
Allow specifying both description and date for tasks

Commit:
Merge branch 'branch-Level-10' into master

Commit:
Fix a bug with the description specifier

Commit:
Add an assertion

Commit:
Improve code quality

zhiayang committed:
Merge pull request #2 from zhiayang/branch-A-CodeQuality

Improve code quality

zhiayang committed:
Merge pull request #1 from zhiayang/branch-A-Assertions

Commit:
Add github actions

Commit:
Add gradle wrapper files

because the gradle people have no idea how a build system
is supposed to work in 2020.

Commit:
Add exec permissions to gradlew

words cannot describe how stupid this is.

Commit:
Change the case on the expected.txt file

Commit:
Attempt to fix gradle failure on windows

Commit:
Perform printf debugging on CI like a fool

Commit:
Make gradle not swallow my output

Commit:
Add gitattributes file

Commit:
Add Github Actions CI

Commit:
Fix checkstyle nags

Commit:
Add editing functionality

Commit:
Update test case after renaming bye -> quit

Commit:
Appease checkstyle again

Commit:
Improve the GUI

zhiayang committed:
Set theme jekyll-theme-minimal

zhiayang committed:
Add ui.png

Commit:
Make the find command search the description as well

Commit:
Add user guide (first draft)

Commit:
trigger-rebuild

zhiayang committed:
Set theme jekyll-theme-minimal

Commit:
Update user guide

Commit:
Merge branch 'master' of github.com:zhiayang/ip into master

zhiayang committed:
Set theme jekyll-theme-midnight

Commit:
Rename userguide file...

Commit:
Try userguide again

Commit:
Try harder

Commit:
why doesn't this work

Commit:
asdf

Commit:
asdf 2

Commit:
asdf 3

Commit:
asdf 4

Commit:
asdf 5

Commit:
asdf 6

Commit:
asdf 7

Commit:
asdf 8

Commit:
asdf 9

Commit:
Fix userguide formatting
